PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/storage/azurite/AzuriteService.kt

```diff
@@ -27,57 +27,92 @@ import com.intellij.execution.process.ColoredProcessHandler
+        private val logger = logger<AzuriteService>()
+    }
+
+    private val sessionLifetimes = SequentialLifetimes(serviceLifetime)
 
-    private val logger = Logger.getInstance(AzuriteService::class.java)
+    private val sessionStarted = AtomicBoolean(false)
+    var session: AzuriteSession = AzuriteNotStartedSession()
+        private set
 
     var processHandler: ColoredProcessHandler? = null
-        get() = field
-        private set(value) {
-            field = value
-        }
+        private set
 
     var workspace: String? = null
-        get() = field
-        private set(value) {
-            field = value
-        }
+        private set
 
     val isRunning: Boolean
         get() {
-            val currentHandler = processHandler
-            return currentHandler != null && !currentHandler.isProcessTerminated
+            return sessionStarted.get() && !sessionLifetimes.isTerminated
         }
 
     fun start(commandLine: GeneralCommandLine, workspaceLocation: String) {
-        if (processHandler != null) {
-            logger.warn("start() was called while processHandler was not null. The caller should verify if an existing session is running, before calling start()")
+        if (isRunning) {
+            logger.warn("The caller should verify if an existing session is running, before calling start()")
             return
         }
 
-        processHandler = object : ColoredProcessHandler(commandLine) {
+        val sessionLifetime = sessionLifetimes.next()
+
+        val newProcessHandler = object : ColoredProcessHandler(commandLine) {
             // If it's a long-running mostly idle daemon process, 'BaseOutputReader.Options.forMostlySilentProcess()' helps to reduce CPU usage.
             override fun readerOptions(): BaseOutputReader.Options = BaseOutputReader.Options.forMostlySilentProcess()
         }
-        workspace = workspaceLocation
-        processHandler?.let {
-            it.addProcessListener(object : ProcessListener {
-                override fun onTextAvailable(p0: ProcessEvent, p1: Key<*>) { }
 
-                override fun processTerminated(e: ProcessEvent) {
-                    processHandler = null
+        sessionLifetime.onTermination {
+            if (!newProcessHandler.isProcessTerminating && !newProcessHandler.isProcessTerminated) {
+                logger.trace("Killing Azurite process")
+                newProcessHandler.killProcess()
+            }
+        }
+
+        newProcessHandler.addProcessListener(object : ProcessListener {
+            override fun onTextAvailable(e: ProcessEvent, outputType: Key<*>) {}
+
+            override fun processTerminated(e: ProcessEvent) {
+                sessionLifetime.executeIfAlive {
+                    logger.trace("Terminating Azurite session lifetime")
+                    sessionLifetime.terminate(true)
                 }
+            }
+
+            override fun startNotified(e: ProcessEvent) {}
+        })
```

Comment:
_(Minor)_ `ProcessAdapter` would help to reduce unnecessary overrides in case you only want to handle one method.
```suggestion
        newProcessHandler.addProcessListener(object : ProcessAdapter() {
            override fun processTerminated(e: ProcessEvent) {
                sessionLifetime.executeIfAlive {
                    logger.trace("Terminating Azurite session lifetime")
                    sessionLifetime.terminate(true)
                }
            }
        })
```